lean4-infoview/src/infoview/rpcInterface.ts

```diff
@@ -0,0 +1,147 @@
+    TaggedText_mapRefs(ci, ct => rs.registerRef(pos, ct.info))
+}
+
+function InfoPopup_registerRefs(rs: RpcSessions, pos: DocumentPosition, ip: InfoPopup): void {
+    if (ip.type) CodeWithInfos_registerRefs(rs, pos, ip.type)
+    if (ip.exprExplicit) CodeWithInfos_registerRefs(rs, pos, ip.exprExplicit)
+}
+
+export async function InteractiveDiagnostics_infoToInteractive(rs: RpcSessions, pos: DocumentPosition, info: InfoWithCtx): Promise<InfoPopup | undefined> {
+    const ret = await rs.call<InfoPopup>(pos, 'Lean.Widget.InteractiveDiagnostics.infoToInteractive', info)
+    if (ret) InfoPopup_registerRefs(rs, pos, ret)
+    return ret
+}
+
+export interface InteractiveHypothesis {
+    names: string[]
+    type: CodeWithInfos
+    val?: CodeWithInfos
+}
+
+export interface InteractiveGoal {
+    hyps: InteractiveHypothesis[]
+    type: CodeWithInfos
+    userName?: string
+}
+
+function InteractiveGoal_registerRefs(rs: RpcSessions, pos: DocumentPosition, g: InteractiveGoal) {
+    CodeWithInfos_registerRefs(rs, pos, g.type)
+    for (const h of g.hyps) {
+        CodeWithInfos_registerRefs(rs, pos, h.type)
+        if (h.val) CodeWithInfos_registerRefs(rs, pos, h.val)
+    }
+}
+
+export interface InteractiveGoals {
+    goals: InteractiveGoal[]
+}
+
+function InteractiveGoals_registerRefs(rs: RpcSessions, pos: DocumentPosition, gs: InteractiveGoals) {
+    for (const g of gs.goals) InteractiveGoal_registerRefs(rs, pos, g)
+}
+
+export async function getInteractiveGoals(rs: RpcSessions, pos: DocumentPosition): Promise<InteractiveGoals | undefined> {
+    const ret = await rs.call<InteractiveGoals>(pos, 'Lean.Widget.getInteractiveGoals', DocumentPosition.toTdpp(pos))
+    if (ret) InteractiveGoals_registerRefs(rs, pos, ret)
+    return ret
+}
+
+export async function getInteractiveTermGoal(rs: RpcSessions, pos: DocumentPosition): Promise<InteractiveGoal | undefined> {
+    const ret = await rs.call<InteractiveGoal>(pos, 'Lean.Widget.getInteractiveTermGoal', DocumentPosition.toTdpp(pos))
+    if (ret) InteractiveGoal_registerRefs(rs, pos, ret)
+    return ret
+}
+
+export type MessageData = RpcPtr<'MessageData'>
+export type MsgEmbed =
+    { expr: CodeWithInfos } |
+    { goal: InteractiveGoal } |
+    { lazyTrace: [number, string, MessageData] }
+
+function MsgEmbed_registerRefs(rs: RpcSessions, pos: DocumentPosition, e: MsgEmbed): void {
+    if ('expr' in e) CodeWithInfos_registerRefs(rs, pos, e.expr)
+    else if ('goal' in e) InteractiveGoal_registerRefs(rs, pos, e.goal)
+    else if ('lazyTrace' in e) rs.registerRef(pos, e.lazyTrace[2])
+}
+
+function TaggedMsg_registerRefs(rs: RpcSessions, pos: DocumentPosition, t: TaggedText<MsgEmbed>): void {
+    const go = (t: TaggedText<MsgEmbed>) => {
+        if ('append' in t) { for (const a of t.append) go(a) }
+        else if ('tag' in t) { MsgEmbed_registerRefs(rs, pos, t.tag[0]); go(t.tag[1]) }
+    }
+    go(t)
+}
```

Comment:
These `registerRefs` functions are 100% boilerplate and are a chore to keep in sync with upstream (and a single typo causes a memory leak).  I think we should just look for `p` attributes and register those.  Then we don't need handcrafted register functions or complicated wrappers (since `RpcSession.call` can register the objects by itself).  I believe you wanted to do it this way originally.  We're also using this approach in the neovim extension and it works great: https://github.com/Julian/lean.nvim/blob/53bd0cc9d5dd27c46f3c1372dc331da4c5372244/lua/lean/rpc.lua#L108-L125

API-wise I think it would be good to include the position in the `RpcContext`.  Right now there's a lot of pointless `pos` arguments everywhere.

Reply:
No, the intention was always to do it explicitly this way. I was (and still am) hoping to auto-generate the code from Lean using a JS backend (see comment at top of file). But there are more important things to work on atm so, although I have reasons to believe it would be significantly easier than an LLVM backend, say, I admit a Lean JS emitter is vaporware. It would also increase the implementation cost for Lua.

Then as a pragmatic matter, I'll agree we can use an ad-hoc traversal; as you mentioned before, `p` is too likely to clash -- if we go with this approach, we should change it to something like `_rpcptr`. This will unfortunately increase payload size.